File: Daily_Challenge/20230329_LC_87.py
class Solution:
    def isScramble(self, s1: str, s2: str) -> bool:
        mem = {}

        def scramble(s1, s2):
            if s1 == s2:
                return True
            if s1 == s2[::-1]:
                return True
            if (s1, s2) in mem:
                return mem[(s1, s2)]
            if Counter(s1) != Counter(s2):
                mem[(s1, s2)] = False
                return False
            mem[(s1, s2)] = False
            l = len(s1)
            for k in range(1, l):
                if scramble(s1[0:k], s2[0:k]) and scramble(s1[k:], s2[k:]):
                    mem[(s1, s2)] = True
                # first and second swap
                if scramble(s1[0:k], s2[l-k:]) and scramble(s1[k:], s2[0:l-k]):
                    mem[(s1, s2)] = True
            return mem[(s1, s2)]
        return scramble(s1, s2)

File: Daily_Challenge/20230329_LC_87.cpp
class Solution {
public:
    vector<int> cal_freq(string s){
        vector<int> ans(26, 0);
        for (char c: s){
            ans[c - 'a']++;
        }
        return ans;
    }
    bool scramble(string s1, string s2, unordered_map<string, unordered_map<string, bool>>& mem){
        if (s1 == s2){
            return true;
        }
        if (mem.count(s1) && mem[s1].count(s2)){
            return mem[s1][s2];
        }
        if (cal_freq(s1) != cal_freq(s2)){
            mem[s1][s2] = false;
            return false;
        }
        mem[s1][s2] = false;
        int L = s1.size();
        for (int i = 1; i < L; i++){
            if (scramble(s1.substr(0, i),s2.substr(0, i), mem) && scramble(s1.substr(i), s2.substr(i), mem)){
                mem[s1][s2] = true;
                return true;
            }
            if (scramble(s1.substr(0, i), s2.substr(L-i, i), mem) && scramble(s1.substr(i), s2.substr(0, L-i), mem)){
                mem[s1][s2] = true;
                return true;
            }
        }
        return mem[s1][s2];
    }
    bool isScramble(string s1, string s2) {
        // without memoization it will TLE
        unordered_map<string, unordered_map<string, bool>> mem;
        return scramble(s1, s2, mem);
    }

};